pootle/apps/pootle_translationproject/models.py

```diff
@@ -424,5 +455,144 @@ def scan_projects(sender, instance, created=False, raw=False, **kwargs):
+        job.save(pipeline=pipe)
+
+    def save_deferred(self, depends_on, pipe):
+        """
+        Preparing job to defer (add as dependent). Works via pipeline.
+        Nothing done if WatchError happens while next `pipeline.execute()`.
+        """
+        job = self.create_job(depends_on=depends_on, status=JobStatus.DEFERRED)
+        job.register_dependency(pipeline=pipe)
+        job.save(pipeline=pipe)
+
+        return job
+
+
+class SyncJobFunctions(object):
+    SYNC = sync_translation_project
+    UPDATE = update_translation_project
+
+
+POOTLE_SYNC_LAST_JOB_PREFIX = "pootle:sync:last:job:"
+
+
+def create_rq_job(language, project, job_func, **options):
+    queue = get_queue('default')
+    queue.connection.sadd(queue.redis_queues_keys, queue.key)
+
+    job_wrapper = JobWrapper.create(job_func,
+                                    language=language,
+                                    project=project,
+                                    options=options,
+                                    connection=queue.connection,
+                                    origin=queue.name,
+                                    timeout=queue.DEFAULT_TIMEOUT)
+
+    last_job_key = job_wrapper.get_last_job_key()
+
+    with queue.connection.pipeline() as pipe:
+        while True:
+            try:
+                pipe.watch(last_job_key)
+                last_job_id = queue.connection.get(last_job_key)
+                depends_on_wrapper = None
+                if last_job_id is not None:
+                    pipe.watch(Job.key_for(last_job_id))
+                    depends_on_wrapper = JobWrapper(last_job_id, queue.connection)
+
+                pipe.multi()
+
+                depends_on_status = None
+                if depends_on_wrapper is not None:
+                    depends_on = depends_on_wrapper.job
+                    depends_on_status = depends_on.get_status()
+
+                if depends_on_status is None:
+                    # enqueue without dependencies
+                    pipe.set(last_job_key, job_wrapper.id)
+                    job_wrapper.save_enqueued(pipe)
+                    pipe.execute()
+                    break
+
+                pipe.set(last_job_key, job_wrapper.id)
+
+                if depends_on_status not in [JobStatus.FINISHED, JobStatus.FAILED]:
+                    # add job as a dependent
+                    job = job_wrapper.save_deferred(last_job_id, pipe)
+                    pipe.execute()
+                    logging.debug('ADD AS DEPENDENT for %s (job_id=%s) OF %s' %
+                                 (last_job_key, job.id, last_job_id))
+                    return job
+
+                job_wrapper.save_enqueued(pipe)
+                pipe.execute()
+                break
+            except WatchError:
+                logging.debug('RETRY after WatchError for %s' % last_job_key)
+                continue
+    logging.debug('ENQUEUE %s (job_id=%s)' % (last_job_key, job_wrapper.id))
```

Comment:
All those logging.debug calls are unnecessary imo, they're just going to be extremely spammy.